src/Dialect/KeywordsDialectProvider.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Behat Gherkin Parser.
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Behat\Gherkin\Dialect;
+
+use Behat\Gherkin\Keywords\KeywordsInterface;
+
+/**
+ * Adapter for the legacy keywords interface.
+ *
+ * @internal
+ */
+final class KeywordsDialectProvider implements DialectProviderInterface
+{
+    public function __construct(
+        private readonly KeywordsInterface $keywords,
+    ) {
+    }
+
+    public function getDialect(string $language): GherkinDialect
+    {
+        // The legacy keywords interface doesn't support detecting whether changing the language worked or no.
```

Comment:
I'm wondering whether we should do a special case by adding an internal `ArrayKeywords::getLanguage` method to read the language, and then using it when the KeywordsInterface implementation extends ArrayKeywords (which is the case for 99.99% of usages at least IMO, as all our core implementations of KeywordsInterface extend it, except the new DialectKeywords internal implementation added in this PR but which has no reason to be wrapped again here except doing crazy things).
On the other hand, not being able to detect the failure here is what preserves the existing behavior, making the switch to the DialectProvider API a kind of opt-in for the strict validation of language comments.
What do you think @acoulton ?

Reply:
I think that's a good idea, and makes all the DialectProvider implementations consistent.

I don't think we can control BC entirely based on which language/dialect provider is injected - I suspect Behat will want to start injecting the new DialectProvider, but that doesn't mean the end-user wants to opt-in to the new strict behaviour?

I'd suggest that the `Lexer->setLanguage()` takes responsibility for the BC based on the GherkinCompatibilityMode - either with something like `$this->dialectProvider->supportsLanguage()`, or by catching the NoSuchLanguageException and switching back to `en` if parsing in legacy mode?

Reply:
I will implement that on Monday.

Reply:
I updated the PR to catch the NoSuchLanguageException to preserve the existing behavior of ignoring failure to change language.
A follow-up PR will enable the strict behavior based on the compatibility mode.